package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -840,82 +457,58 @@ cdef class FastNS(object):
         if any of the query coordinates lies outside the `box` supplied to
         :class:`~MDAnalysis.lib.nsgrid.FastNS`.
         """
-
-        cdef ns_int i, j, size_search
-        cdef ns_int d, m
-        cdef ns_int current_beadid, bid
-        cdef ns_int cellindex, cellindex_probe
-        cdef ns_int xi, yi, zi
-
+        from MDAnalysis.lib import distances
+
+        cdef int i, j, size_search
+        cdef int cx, cy, cz
+        cdef int cellid
+        cdef int xi, yi, zi
+        cdef int cellcoord[3], searchcoord[3]
+        
         cdef NSResults results
-
-        cdef dreal d2
-        cdef rvec probe
-
-        cdef real[:, ::1] searchcoords
-        cdef real[:, ::1] searchcoords_bbox
-        cdef _NSGrid searchgrid
-        cdef bint check
-
-        cdef dreal cutoff2 = self.cutoff * self.cutoff
-        cdef ns_int npairs = 0
+        cdef double d2, cutoff2
+        cdef float[:, ::1] searchcoords_bbox
+        
+        cutoff2 = self.cutoff * self.cutoff
 
         if (search_coords.ndim != 2 or search_coords.shape[1] != 3):
             raise ValueError("search_coords must have a shape of (n, 3), got "
                              "{}.".format(search_coords.shape))
 
-        # Generate another grid to search
-        searchcoords = search_coords.astype(np.float32, order='C', copy=False)
-        searchcoords_bbox = self.box.fast_put_atoms_in_bbox(searchcoords)
-        searchgrid = _NSGrid(searchcoords_bbox.shape[0], self.grid.used_cutoff, self.box, self.max_gridsize, force=True)
-
-        size_search = searchcoords.shape[0]
+        searchcoords_bbox =  distances.apply_PBC(search_coords, self.grid.dimensions)
+        size_search = search_coords.shape[0]
+        results = NSResults()
+
+        for i in range(size_search):
+            # which cell is atom *i* in
+            self.grid.coord2cellxyz(&searchcoords_bbox[i][0], &cellcoord[0])
+            # loop over all 27 neighbouring cells
+            for xi in range(3):
+                for yi in range(3):
+                    for zi in range(3):
+                        cx = cellcoord[0] - 1 + xi
+                        cy = cellcoord[1] - 1 + yi
+                        cz = cellcoord[2] - 1 + zi
+                        cellid = self.grid.cellxyz2cellid(cx, cy, cz)
+
+                        if not self.periodic and cellid == -1:  # out of bounds
+                            continue
+
+                        # for loop over atoms in searchcoord
+                        j = self.grid.head_id[cellid]
```

Comment:
This is where some of the tests (the triclinic ones) currently segfault. The reason is that `coord2cellxyz()` is called on line 484, which may yield an invalid `cellcoord` for triclinic boxes. Since `cellxyz2cellid()` (called on line 492) only checks if the coordinates are *equal to* (and not equal or greater than) `ncells`, the returned `cellid` can be out of bounds, causing a segfault.